SceneAlgo::hierarchyHash

danieldresser-ie opened the pull request:
I'm feeling fairly good about this overall, and how it fits in with what we're planning for Instancer.

I don't really like the current structure of this code though - my initial implementation was pretty clean, but it incorrectly included hashes of full path, not relative paths ( which would cause problems when we want to use this for Instancers with relative prototypes ). To fix that, I really want to do a parallel reduce, similar to what you needed in USDLayerWriter. I realized I can actually do a reduce using the way that functors are allocated by parallelProcessLocations, which is sort of neat, but this workaround is not good for clarity. If we're not currently adding a parallelGatherLocations, I should probably just implement my own walk function, same as is currently done in USDLayerWriter.

The constraint that we want identical subhierarchies to hash the same is an interesting one - we require different valeus to have different hashes, but we don't generally require the same value to have the same hash. I suspect there are a lot of ways of constructing identical hierarchies where their hashes would end up being different, but this code does now seem to be working when the hashes of each element are consistent. ( When I first tested it, I was testing on the root location, and then tried to get the hash of a variety of stuff at the root to match the same stuff under a group ... but I wasn't able to get it to match. I suspect this is because evaluating the `objectPlug` at the root is a special case, and its hash will never match another location - even if that location doesn't have any object either. Anyway, for now I've just changed the test to compare two groups, one of which is nested an extra layer deep, and those do have matching hashes. )

johnhaddon commented on include/GafferScene/SceneAlgo.h:
```diff
@@ -150,6 +150,29 @@ void parallelGatherLocations(
+template <typename T, typename LocationFunctor, typename ReduceFunctor>
+T parallelReduceLocations(
+	const ScenePlug *scene,
+	const T& identity,
```
(I realise we'd need to use `invoke_result_t` for this, but I think that's OK?

danieldresser-ie replied:
As discussed this morning - the main reason to expose this is actually not because we expect people to want to use it very often, but because it communicates some of the otherwise unexpected things that could happen - using an Imath color would result in uninitialized memory with a default `T()`. But also with a custom type, it's important to make sure that the value provided is actually an identity, because accumulators will be initialized to this value an unknown number of times - if the default constructed value has any effect when merged with something, you will also get a non-deterministic result. This value as a signpost for weird behaviour is reduced if the signature encourages you to ignore passing it, so I'm kind of inclined to leave this where it is.

johnhaddon reviewed (commented):
Thanks Daniel! The SceneAlgo stuff is feeling pretty good to me - a few comments inline, but not much. Your comments indicate that the performance test isn't any good - would be good to sort that out. USDLayerWriter stuff also seems like it needs a bit of tidying up, and benchmarking against the original - could perhaps do that in a separate PR?

danieldresser-ie commented on include/GafferScene/SceneAlgo.h:
```diff
@@ -292,6 +315,14 @@ GAFFERSCENE_API IECore::PathMatcher linkedLights( const ScenePlug *scene, const
+
+// Hashes all properties of a location and all its children. Includes the names of the children relative to
+// the given location ( but not their order - two locations will hash the same if the only difference is the order
+// of their children ). Does not include set membership.
```
Fixed.

danieldresser-ie commented on include/GafferScene/SceneAlgo.inl:
```diff
@@ -287,4 +288,161 @@ IECore::PathMatcher findAll( const ScenePlug *scene, Predicate &&predicate, cons
+	using ChildNameRange = tbb::blocked_range<std::vector<IECore::InternedString>::const_iterator>;
+	const ChildNameRange loopRange( childNames.begin(), childNames.end() );
+
+	if( childNames.empty() )
```
Fixed.

danieldresser-ie commented on include/GafferScene/SceneAlgo.inl:
```diff
@@ -287,4 +288,161 @@ IECore::PathMatcher findAll( const ScenePlug *scene, Predicate &&predicate, cons
+		const T& m_identity;
+		LocationFunctor &&m_locationFunctor;
+		MergeChildrenFunctor &&m_mergeChildrenFunctor;
+		ReduceFunctor &&m_reduceFunctor;
+		tbb::task_group_context &m_taskGroupContext;
+
+		LoopBody(
+			const GafferScene::ScenePlug *scene,
+			const Gaffer::ThreadState &threadState,
+			const ScenePlug::ScenePath &path,
+			const T& identity,
+			LocationFunctor &&locationFunctor,
+			MergeChildrenFunctor &&mergeChildrenFunctor,
+			ReduceFunctor &&reduceFunctor,
+			tbb::task_group_context &taskGroupContext
+		) :
+			m_result( identity ),
+			m_scene( scene ),
+			m_threadState( threadState ),
+			m_path( path ),
+			m_identity( identity ),
+			m_locationFunctor( locationFunctor ),
+			m_mergeChildrenFunctor( mergeChildrenFunctor ),
+			m_reduceFunctor( reduceFunctor ),
+			m_taskGroupContext( taskGroupContext )
+		{
+		}
+
+		LoopBody( LoopBody& s, tbb::split ) :
+			m_result( s.m_identity ),
+			m_scene( s.m_scene ),
+			m_threadState( s.m_threadState ),
+			m_path( s.m_path ),
+			m_identity( s.m_identity ),
+			m_locationFunctor( s.m_locationFunctor ),
+			m_mergeChildrenFunctor( s.m_mergeChildrenFunctor ),
+			m_reduceFunctor( s.m_reduceFunctor ),
+			m_taskGroupContext( s.m_taskGroupContext )
+		{
+		}
```
Added the comment that I was somehow sure I had previously written ( but apparently wrote only in my mind ) :

```
    // It's a shame to be using all this boilerplate which could be avoided with a [&] in a lambda,
    // but using the imperative form of parallel_reduce allows us avoid unnecessary allocations.
    // This internal ugliness is probably worthwhile if it gives us a public interface and functionality
    // we're happy with.
```

danieldresser-ie commented on include/GafferScene/SceneAlgo.h:
```diff
@@ -292,6 +315,14 @@ GAFFERSCENE_API IECore::PathMatcher linkedLights( const ScenePlug *scene, const
+// Hashes all properties of a location and all its children. Includes the names of the children relative to
+// the given location ( but not their order - two locations will hash the same if the only difference is the order
+// of their children ). Does not include set membership.
+GAFFERSCENE_API IECore::MurmurHash hierarchyHash( const ScenePlug *scene, const ScenePlug::ScenePath &location );
```
Renamed.

danieldresser-ie commented:
OK, I think I've now addressed all feedback, and added Changelog entries.

I've also added a comment to the test, and I'll clarify a bit more here what I meant by the perf test being "not very good":
It's just very dominated by the time to actually hash the plugs, so none of the rest of what it's doing is very noticeable - when I was looking at the old approach, and comparing an atomic to thread local storage, neither made any perceptible difference. In particular, the sort of thing I would like this test to show is that it's the right decision to use parallel_reduce instead of parallel_deterministic_reduce ... it completely fails to show that, there is absolutely no difference. It still feels like a reasonable decision to use parallel_reduce ( using simple adds to combine hashes in an order-independent way is an approach we're relying on elsewhere, so it should be fine here, and USDLayerWriter doesn't rely on order ) ... but I haven't really been able to prove any concrete benefit.

danieldresser-ie commented:
I should flag one last change: I've also added an extra test for hierarchyHash that tries to ensure the internal parallelReduceLocations mechanism is functioning properly.

johnhaddon commented:
Thanks Daniel!
